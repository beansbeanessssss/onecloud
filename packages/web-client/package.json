{
  "name": "@ownclouders/web-client",
  "version": "0.0.1",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/generated && docker run --rm -v \"${PWD}/src:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated"
  },
  "peerDependencies": {
    "@ownclouders/web-client": "workspace:*",
    "@casl/ability": "^6.3.3",
    "axios": "1.6.8",
    "fast-xml-parser": "4.3.3",
    "lodash-es": "^4.17.21",
    "luxon": "3.2.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "webdav": "5.3.1",
    "xml-js": "^1.6.11",
    "zod": "3.22.4"
  }
}
