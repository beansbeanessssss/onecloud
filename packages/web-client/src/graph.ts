import { AxiosInstance, AxiosPromise, AxiosResponse } from 'axios'
import {
  CollectionOfDrives,
  CollectionOfGroup,
  CollectionOfUser,
  Configuration,
  Drive,
  DrivesApiFactory,
  Group,
  GroupApiFactory,
  GroupsApiFactory,
  MeChangepasswordApiFactory,
  MeDrivesApi,
  MeUserApiFactory,
  User,
  UserApiFactory,
  UsersApiFactory,
  TagsApiFactory,
  CollectionOfTags,
  TagAssignment,
  TagUnassignment,
  DrivesGetDrivesApi,
  CollectionOfApplications,
  ApplicationsApiFactory,
  UserAppRoleAssignmentApiFactory,
  AppRoleAssignment,
  ExportPersonalDataRequest,
  MeDriveApiFactory,
  RoleManagementApiFactory,
  UnifiedRoleDefinition,
  CollectionOfDriveItems1,
  DriveItemApiFactory,
  DrivesRootApiFactory,
  DriveItem,
  DrivesPermissionsApiFactory,
  Permission,
  DriveItemCreateLink,
  DriveItemInvite,
  SharingLinkPassword
} from './generated'

export interface Graph {
  applications: {
    listApplications: () => AxiosPromise<CollectionOfApplications>
  }
  tags: {
    getTags: () => AxiosPromise<CollectionOfTags>
    assignTags: (tagAssignment?: TagAssignment) => AxiosPromise<void>
    unassignTags: (tagUnassignment?: TagUnassignment) => AxiosPromise<void>
  }
  drives: {
    listMyDrives: (orderBy?: string, filter?: string) => Promise<AxiosResponse<CollectionOfDrives>>
    listAllDrives: (orderBy?: string, filter?: string) => Promise<AxiosResponse<CollectionOfDrives>>
    listSharedWithMe: () => AxiosPromise<CollectionOfDriveItems1>
    listSharedByMe: () => AxiosPromise<CollectionOfDriveItems1>
    getDrive: (id: string) => AxiosPromise<Drive>
    createDrive: (drive: Drive, options: any) => AxiosPromise<Drive>
    updateDrive: (id: string, drive: Drive, options: any) => AxiosPromise<Drive>
    disableDrive: (id: string) => AxiosPromise<void>
    deleteDrive: (id: string) => AxiosPromise<void>
    deleteDriveItem: (driveId: string, itemId: string) => AxiosPromise<void>
    createDriveItem: (driveId: string, driveItem: DriveItem) => AxiosPromise<DriveItem>
  }
  users: {
    getUser: (userId: string) => AxiosPromise<User>
    createUser: (user: User) => AxiosPromise<User>
    getMe: () => AxiosPromise<User>
    editMe: (user: User) => AxiosPromise<User>
    changeOwnPassword: (currentPassword: string, newPassword: string) => AxiosPromise<void>
    editUser: (userId: string, user: User) => AxiosPromise<User>
    deleteUser: (userId: string) => AxiosPromise<void>
    listUsers: (
      orderBy?: string,
      filter?: string,
      expand?: Array<'drive' | 'drives' | 'memberOf' | 'appRoleAssignments'>,
      search?: string
    ) => AxiosPromise<CollectionOfUser>
    createUserAppRoleAssignment: (
      userId: string,
      appRoleAssignment: AppRoleAssignment
    ) => AxiosPromise<AppRoleAssignment>
    exportPersonalData: (
      userId: string,
      exportPersonalDataRequest?: ExportPersonalDataRequest
    ) => AxiosPromise<void>
  }
  groups: {
    listGroups: (
      orderBy?: string,
      expand?: Array<'members'>,
      search?: string
    ) => AxiosPromise<CollectionOfGroup>
    createGroup: (group: Group) => AxiosPromise<Group>
    getGroup: (groupId: string) => AxiosPromise<Group>
    editGroup: (groupId: string, group: Group) => AxiosPromise<void>
    deleteGroup: (groupId: string) => AxiosPromise<void>
    addMember: (groupId: string, userId: string, server: string) => AxiosPromise<void>
    deleteMember: (groupId: string, userId: string) => AxiosPromise<void>
  }
  roleManagement: {
    listPermissionRoleDefinitions: () => AxiosPromise<UnifiedRoleDefinition>
  }
  permissions: {
    getPermission: (driveId: string, itemId: string, permId: string) => AxiosPromise<Permission>
    listPermissions: (driveId: string, itemId: string) => AxiosPromise<Permission[]>
    createLink: (
      driveId: string,
      itemId: string,
      driveItemCreateLink?: DriveItemCreateLink
    ) => AxiosPromise<Permission>
    invite: (
      driveId: string,
      itemId: string,
      driveItemInvite?: DriveItemInvite
    ) => AxiosPromise<Permission>
    deletePermission: (driveId: string, itemId: string, permId: string) => AxiosPromise<void>
    updatePermission: (
      driveId: string,
      itemId: string,
      permId: string,
      permission: Permission
    ) => AxiosPromise<Permission>
    setPermissionPassword: (
      driveId: string,
      itemId: string,
      permId: string,
      sharingLinkPassword: SharingLinkPassword
    ) => AxiosPromise<Permission>
  }
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  const meDrivesApi = new MeDrivesApi(config, config.basePath, axiosClient)
  const allDrivesApi = new DrivesGetDrivesApi(config, config.basePath, axiosClient)
  const meUserApiFactory = MeUserApiFactory(config, config.basePath, axiosClient)
  const meDriveApiFactory = MeDriveApiFactory(config, config.basePath, axiosClient)
  const meChangepasswordApiFactory = MeChangepasswordApiFactory(
    config,
    config.basePath,
    axiosClient
  )
  const applicationsApiFactory = ApplicationsApiFactory(config, config.basePath, axiosClient)
  const userApiFactory = UserApiFactory(config, config.basePath, axiosClient)
  const usersApiFactory = UsersApiFactory(config, config.basePath, axiosClient)
  const userAppRoleAssignmentApiFactory = UserAppRoleAssignmentApiFactory(
    config,
    config.basePath,
    axiosClient
  )
  const groupApiFactory = GroupApiFactory(config, config.basePath, axiosClient)
  const groupsApiFactory = GroupsApiFactory(config, config.basePath, axiosClient)
  const drivesApiFactory = DrivesApiFactory(config, config.basePath, axiosClient)
  const tagsApiFactory = TagsApiFactory(config, config.basePath, axiosClient)
  const roleManagementApiFactory = RoleManagementApiFactory(config, config.basePath, axiosClient)
  const driveItemApiFactory = DriveItemApiFactory(config, config.basePath, axiosClient)
  const drivesRootApiFactory = DrivesRootApiFactory(config, config.basePath, axiosClient)
  const drivesPermissionsApiFactory = DrivesPermissionsApiFactory(
    config,
    config.basePath,
    axiosClient
  )

  return <Graph>{
    applications: {
      listApplications: () => applicationsApiFactory.listApplications()
    },
    tags: {
      getTags: () => tagsApiFactory.getTags(),
      assignTags: (tagAssignment: TagAssignment) => tagsApiFactory.assignTags(tagAssignment),
      unassignTags: (tagUnassignment: TagUnassignment) =>
        tagsApiFactory.unassignTags(tagUnassignment)
    },
    drives: {
      listMyDrives: (orderBy?: string, filter?: string) =>
        meDrivesApi.listMyDrives(orderBy, filter),
      listAllDrives: (orderBy?: string, filter?: string) =>
        allDrivesApi.listAllDrives(orderBy, filter),
      listSharedWithMe: () => meDriveApiFactory.listSharedWithMe(),
      listSharedByMe: () => meDriveApiFactory.listSharedByMe(),
      getDrive: (id: string) => drivesApiFactory.getDrive(id),
      createDrive: (drive: Drive, options: any): AxiosPromise<Drive> =>
        drivesApiFactory.createDrive(drive, options),
      updateDrive: (id: string, drive: Drive, options: any): AxiosPromise<Drive> =>
        drivesApiFactory.updateDrive(id, drive, options),
      disableDrive: (id: string): AxiosPromise<void> => drivesApiFactory.deleteDrive(id, '', {}),
      deleteDrive: (id: string): AxiosPromise<void> =>
        drivesApiFactory.deleteDrive(id, '', {
          headers: {
            Purge: 'T'
          }
        }),
      deleteDriveItem: (driveId: string, itemId: string) =>
        driveItemApiFactory.deleteDriveItem(driveId, itemId),
      createDriveItem: (driveId: string, driveItem: DriveItem) =>
        drivesRootApiFactory.createDriveItem(driveId, driveItem)
    },
    users: {
      getUser: (userId: string) =>
        userApiFactory.getUser(
          userId,
          new Set<any>([]),
          new Set<any>(['drive', 'memberOf', 'appRoleAssignments'])
        ),
      createUser: (user: User) => usersApiFactory.createUser(user),
      getMe: () => meUserApiFactory.getOwnUser(new Set<any>(['memberOf'])),
      editMe: (user: User) => meUserApiFactory.updateOwnUser(user),
      changeOwnPassword: (currentPassword, newPassword) =>
        meChangepasswordApiFactory.changeOwnPassword({ currentPassword, newPassword }),
      editUser: (userId: string, user: User) => userApiFactory.updateUser(userId, user),
      deleteUser: (userId: string) => userApiFactory.deleteUser(userId),
      listUsers: (
        orderBy?: any,
        filter?: string,
        expand?: Array<'drive' | 'drives' | 'memberOf' | 'appRoleAssignments'>,
        search: string = ''
      ) =>
        usersApiFactory.listUsers(
          search,
          filter,
          orderBy ? new Set<any>([orderBy]) : null,
          null,
          expand ? new Set<any>(expand) : null
        ),
      createUserAppRoleAssignment: (userId: string, appRoleAssignment: AppRoleAssignment) =>
        userAppRoleAssignmentApiFactory.userCreateAppRoleAssignments(userId, appRoleAssignment),
      exportPersonalData: (userId: string, exportPersonalDataRequest?: ExportPersonalDataRequest) =>
        userApiFactory.exportPersonalData(userId, exportPersonalDataRequest)
    },
    groups: {
      createGroup: (group: Group) => groupsApiFactory.createGroup(group),
      editGroup: (groupId: string, group: Group) => groupApiFactory.updateGroup(groupId, group),
      getGroup: (groupId: string) =>
        groupApiFactory.getGroup(groupId, new Set<any>([]), new Set<any>(['members'])),
      deleteGroup: (groupId: string) => groupApiFactory.deleteGroup(groupId),
      listGroups: (orderBy?: any, expand?: Array<'members'>, search: string = '') =>
        groupsApiFactory.listGroups(
          search,
          orderBy ? new Set<any>([orderBy]) : null,
          null,
          expand ? new Set<any>(expand) : null
        ),
      addMember: (groupId: string, userId: string, server: string) =>
        groupApiFactory.addMember(groupId, { '@odata.id': `${server}graph/v1.0/users/${userId}` }),
      deleteMember: (groupId: string, userId: string) =>
        groupApiFactory.deleteMember(groupId, userId)
    },
    roleManagement: {
      listPermissionRoleDefinitions: () => roleManagementApiFactory.listPermissionRoleDefinitions()
    },
    permissions: {
      getPermission: (driveId: string, itemId: string, permId: string) =>
        drivesPermissionsApiFactory.getPermission(driveId, itemId, permId),
      listPermissions: (driveId: string, itemId: string) =>
        drivesPermissionsApiFactory.listPermissions(driveId, itemId),
      createLink: (driveId: string, itemId: string, driveItemCreateLink?: DriveItemCreateLink) =>
        drivesPermissionsApiFactory.createLink(driveId, itemId, driveItemCreateLink),
      invite: (driveId: string, itemId: string, driveItemInvite?: DriveItemInvite) =>
        drivesPermissionsApiFactory.invite(driveId, itemId, driveItemInvite),
      deletePermission: (driveId: string, itemId: string, permId: string) =>
        drivesPermissionsApiFactory.deletePermission(driveId, itemId, permId),
      updatePermission: (driveId: string, itemId: string, permId: string, permission: Permission) =>
        drivesPermissionsApiFactory.updatePermission(driveId, itemId, permId, permission),
      setPermissionPassword: (
        driveId: string,
        itemId: string,
        permId: string,
        sharingLinkPassword: SharingLinkPassword
      ) =>
        drivesPermissionsApiFactory.setPermissionPassword(
          driveId,
          itemId,
          permId,
          sharingLinkPassword
        )
    }
  }
}
