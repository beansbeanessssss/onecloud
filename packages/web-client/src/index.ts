import axios, { AxiosInstance, AxiosPromise, AxiosResponse } from 'axios'
import {
  Configuration,
  MeDrivesApi,
  Drive,
  DrivesApiFactory,
  CollectionOfDrives,
  UserApiFactory,
  User,
  MeUserApiFactory,
  UsersApiFactory,
  GroupsApiFactory,
  MeChangepasswordApiFactory,
  Group,
  CollectionOfGroup,
  CollectionOfUser,
  GroupApiFactory
} from './generated'

export interface Graph {
  drives: {
    listMyDrives: (orderBy?: string, filter?: string) => Promise<AxiosResponse<CollectionOfDrives>>
    getDrive: (id: string) => AxiosPromise<Drive>
    createDrive: (drive: Drive, options: any) => AxiosPromise<Drive>
    updateDrive: (id: string, drive: Drive, options: any) => AxiosPromise<Drive>
    deleteDrive: (id: string, ifMatch: string, options: any) => AxiosPromise<void>
  }
  users: {
    getUser: (userId: string) => AxiosPromise<User>
    createUser: (user: User) => AxiosPromise<User>
    getMe: () => AxiosPromise<User>
    changeOwnPassword: (currentPassword: string, newPassword: string) => AxiosPromise<void>
    editUser: (userId: string, user: User) => AxiosPromise<User>
    deleteUser: (userId: string) => AxiosPromise<void>
    listUsers: (orderBy?: string) => AxiosPromise<CollectionOfUser>
  }
  groups: {
    listGroups: (orderBy?: string) => AxiosPromise<CollectionOfGroup>
    createGroup: (group: Group) => AxiosPromise<Group>
    editGroup: (groupId: string, group: Group) => AxiosPromise<void>
    deleteGroup: (groupId: string) => AxiosPromise<void>
    addMember: (groupId: string, userId: string, server: string) => AxiosPromise<void>
    deleteMember: (groupId: string, userId: string) => AxiosPromise<void>
  }
}

const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const basePath = new URL('/graph/v1.0', baseURI).href
  const config = new Configuration({
    basePath
  })

  const meDrivesApi = new MeDrivesApi(config, config.basePath, axiosClient)
  const meUserApiFactory = MeUserApiFactory(config, config.basePath, axiosClient)
  const meChangepasswordApi = MeChangepasswordApiFactory(config, config.basePath, axiosClient)
  const userApiFactory = UserApiFactory(config, config.basePath, axiosClient)
  const usersApiFactory = UsersApiFactory(config, config.basePath, axiosClient)
  const groupApiFactory = GroupApiFactory(config, config.basePath, axiosClient)
  const groupsApiFactory = GroupsApiFactory(config, config.basePath, axiosClient)
  const drivesApiFactory = DrivesApiFactory(config, config.basePath, axiosClient)

  return <Graph>{
    drives: {
      listMyDrives: (orderBy?: string, filter?: string) =>
        meDrivesApi.listMyDrives(0, 0, orderBy, filter),
      getDrive: (id: string) => drivesApiFactory.getDrive(id),
      createDrive: (drive: Drive, options: any): AxiosPromise<Drive> =>
        drivesApiFactory.createDrive(drive, options),
      updateDrive: (id: string, drive: Drive, options: any): AxiosPromise<Drive> =>
        drivesApiFactory.updateDrive(id, drive, options),
      deleteDrive: (id: string, ifMatch: string, options: any): AxiosPromise<void> =>
        drivesApiFactory.deleteDrive(id, ifMatch, options)
    },
    users: {
      getUser: (userId: string) => userApiFactory.getUser(userId),
      createUser: (user: User) => usersApiFactory.createUser(user),
      getMe: () => meUserApiFactory.meGet(),
      changeOwnPassword: (currentPassword, newPassword) =>
        meChangepasswordApi.changeOwnPassword({ currentPassword, newPassword }),
      editUser: (userId: string, user: User) => userApiFactory.updateUser(userId, user),
      deleteUser: (userId: string) => userApiFactory.deleteUser(userId),
      listUsers: (orderBy?: any) =>
        usersApiFactory.listUsers(
          0,
          0,
          '',
          '',
          false,
          new Set<any>([orderBy]),
          new Set<any>([]),
          new Set<any>(['memberOf'])
        )
    },
    groups: {
      createGroup: (group: Group) => groupsApiFactory.createGroup(group),
      editGroup: (groupId: string, group: Group) => groupApiFactory.updateGroup(groupId, group),
      deleteGroup: (groupId: string) => groupApiFactory.deleteGroup(groupId),
      listGroups: (orderBy?: any) =>
        groupsApiFactory.listGroups(
          0,
          0,
          '',
          '',
          false,
          new Set<any>([orderBy]),
          new Set<any>([]),
          new Set<any>(['members'])
        ),
      addMember: (groupId: string, userId: string, server: string) => {
        groupApiFactory.addMember(groupId, { '@odata.id': `${server}graph/v1.0/users/${userId}` })
      },
      deleteMember: (groupId: string, userId: string) => {
        groupApiFactory.deleteMember(groupId, userId)
      }
    }
  }
}

interface Client {
  graph: Graph
}

export const client = (baseURI: string, token: string): Client => {
  const axiosClient = axios.create({
    headers: {
      authorization: 'Bearer ' + token,
      'Content-Type': 'application/x-www-form-urlencoded'
    }
  })

  axiosClient.interceptors.request.use((config) => {
    config.headers['Content-Type'] = 'application/x-www-form-urlencoded'
    return config
  })

  return {
    graph: graph(baseURI, axiosClient)
  }
}
