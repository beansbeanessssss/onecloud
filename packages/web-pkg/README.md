# web-pkg

This is experimental.
