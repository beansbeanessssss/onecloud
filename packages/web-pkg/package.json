{
  "name": "@ownclouders/web-pkg",
  "version": "0.0.0",
  "description": "ownCloud web pkg",
  "license": "AGPL-3.0",
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "^0.27.2",
    "filesize": "^9.0.11",
    "fuse.js": "^6.5.3",
    "lodash-es": "^4.17.21",
    "luxon": "^2.4.0",
    "mark.js": "^8.11.1",
    "popper-max-size-modifier": "^0.2.0",
    "qs": "^6.10.3",
    "uuid": "^9.0.0",
    "vue-router": "4.1.6",
    "vuex": "4.1.0",
    "web-client": "npm:@ownclouders/web-client",
    "web-pkg": "npm:@ownclouders/web-pkg",
    "vue-concurrency": "4.0.0"
  }
}
