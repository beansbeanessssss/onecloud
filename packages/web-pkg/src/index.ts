export * from './components'
export * from './composables'
export * from './configuration'
export * from './errors'
export * from './helpers'
export * from './http'
export * from './observer'
export * from './services'
export * from './utils'
