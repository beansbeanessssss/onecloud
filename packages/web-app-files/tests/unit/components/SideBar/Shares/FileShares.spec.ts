import FileShares from 'web-app-files/src/components/SideBar/Shares/FileShares.vue'
import { mock } from 'vitest-mock-extended'
import { Resource } from '@ownclouders/web-client'
import { SpaceResource } from '@ownclouders/web-client'
import { v4 as uuidV4 } from 'uuid'
import { CollaboratorShare, ShareRole, ShareTypes } from '@ownclouders/web-client'
import {
  defaultPlugins,
  mount,
  shallowMount,
  defaultComponentMocks,
  defaultStubs
} from 'web-test-helpers'
import CollaboratorListItem from '../../../../../src/components/SideBar/Shares/Collaborators/ListItem.vue'
import { CapabilityStore, useCanShare, useModals } from '@ownclouders/web-pkg'
import { User } from '@ownclouders/web-client/graph/generated'

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useCanShare: vi.fn()
}))

const getCollaborator = (): CollaboratorShare => ({
  id: uuidV4(),
  sharedWith: {
    id: 'einstein',
    displayName: 'Albert Einstein'
  },
  sharedBy: {
    id: 'admin',
    displayName: 'Admin'
  },
  permissions: [],
  role: mock<ShareRole>(),
  resourceId: uuidV4(),
  indirect: false,
  shareType: ShareTypes.user.value
})

describe('FileShares', () => {
  describe('invite collaborator form', () => {
    it('renders the form when the resource can be shared', () => {
      const resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => true })
      const { wrapper } = getWrapper({ resource, canShare: true })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
    it('does not render the form when the resource can not be shared', () => {
      const resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => false })
      const { wrapper } = getWrapper({ resource, canShare: false })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
    it('does render the form when the resource is a received share and re-sharing is enabled', () => {
      const resource = mock<Resource>({ isReceivedShare: () => true, canShare: () => true })
      const { wrapper } = getWrapper({ resource })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
  })

  describe('collaborators list', () => {
    let collaborators
    beforeEach(() => {
      collaborators = [getCollaborator(), getCollaborator(), getCollaborator(), getCollaborator()]
    })

    it('renders sharedWithLabel and sharee list', () => {
      const { wrapper } = getWrapper({ collaborators })
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(collaborators.length)
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('reacts on delete events', async () => {
      const { wrapper } = getWrapper({ collaborators })
      const { dispatchModal } = useModals()
      wrapper
        .findComponent<typeof CollaboratorListItem>('collaborator-list-item-stub')
        .vm.$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(dispatchModal).toHaveBeenCalledTimes(1)
    })
    it('correctly passes the shared parent route to the collaborator list item for indirect shares', () => {
      const indirectCollaborator = { ...getCollaborator(), indirect: true }
      const ancestorMetaData = {
        '/somePath': { id: indirectCollaborator.resourceId }
      }
      const { wrapper } = getWrapper({ collaborators: [indirectCollaborator], ancestorMetaData })
      const listItemStub = wrapper.findComponent<typeof CollaboratorListItem>(
        'collaborator-list-item-stub'
      )
      expect(listItemStub.props('sharedParentRoute')).toBeTruthy()
      expect(listItemStub.props('modifiable')).toBeFalsy()
    })
    it('toggles the share list', async () => {
      const showAllOnLoad = true
      const { wrapper } = getWrapper({ mountType: mount, collaborators })
      expect(wrapper.vm.sharesListCollapsed).toBe(!showAllOnLoad)
      await wrapper.find('.toggle-shares-list-btn').trigger('click')
      expect(wrapper.vm.sharesListCollapsed).toBe(showAllOnLoad)
    })
    it('share should be modifiable if its personal space share', () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(true)
    })
    it('share should not be modifiable if its not personal space share', () => {
      const space = mock<SpaceResource>({ driveType: 'project' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should not be modifiable if collaborator is indirect', () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      collaborators[0]['indirect'] = true
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should not be modifiable if user is not manager', () => {
      const space = mock<SpaceResource>({ driveType: 'project' })
      ;(space as any).isManager = vi.fn(() => false)
      collaborators[0]['indirect'] = true
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
  })

  describe('current space', () => {
    it('loads space members if a space is given and the current user is member', () => {
      const user = { id: '1' } as User
      const space = mock<SpaceResource>({ driveType: 'project' })
      const spaceMembers = [
        { sharedWith: { id: user.id } },
        { sharedWith: { id: '2' } }
      ] as CollaboratorShare[]
      const collaborator = getCollaborator()
      collaborator.sharedWith = {
        ...collaborator.sharedWith,
        id: user.id
      }
      const { wrapper } = getWrapper({ space, collaborators: [collaborator], user, spaceMembers })
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(1)
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('does not load space members if a space is given but the current user not a member', () => {
      const user = { id: '1' } as User
      const space = mock<SpaceResource>({ driveType: 'project' })
      const spaceMembers = [{ collaborator: { name: `${user}-2` } }]
      const collaborator = getCollaborator()
      collaborator.sharedWith = {
        ...collaborator.sharedWith,
        id: user.id
      }
      const { wrapper } = getWrapper({ space, collaborators: [collaborator], user, spaceMembers })
      expect(wrapper.find('#space-collaborators-list').exists()).toBeFalsy()
    })
  })

  describe('"deleteShareConfirmation" method', () => {
    it('dispatches a modal', () => {
      const { wrapper } = getWrapper()
      const { dispatchModal } = useModals()
      wrapper.vm.deleteShareConfirmation(mock<CollaboratorShare>())
      expect(dispatchModal).toHaveBeenCalled()
    })
  })
})

function getWrapper({
  mountType = shallowMount,
  resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => true }),
  space = mock<SpaceResource>(),
  collaborators = [],
  spaceMembers = [],
  user = undefined,
  showAllOnLoad = true,
  ancestorMetaData = {},
  canShare = true
} = {}) {
  vi.mocked(useCanShare).mockReturnValue({ canShare: () => canShare })

  const capabilities = {
    files_sharing: { deny_access: false }
  } satisfies Partial<CapabilityStore['capabilities']>

  return {
    wrapper: mountType(FileShares, {
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              userState: { user },
              spacesState: { spaceMembers },
              capabilityState: { capabilities },
              configState: {
                options: { contextHelpers: true, sidebar: { shares: { showAllOnLoad } } }
              },
              sharesState: { collaboratorShares: collaborators },
              resourcesStore: { ancestorMetaData }
            }
          })
        ],
        mocks: defaultComponentMocks(),
        provide: { resource, space },
        stubs: {
          ...defaultStubs,
          OcButton: false,
          'invite-collaborator-form': true,
          'collaborator-list-item': true
        }
      }
    })
  }
}
