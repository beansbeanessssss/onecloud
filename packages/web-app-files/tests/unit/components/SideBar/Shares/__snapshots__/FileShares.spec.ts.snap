// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileShares collaborators list correctly passes the shared parent route to the collaborator list item 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm">Share with people</h3>
    <oc-contextual-helper-stub title="Invite specific people" text="Enter a name or group to share this item." list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" endtext="" readmorelink="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" class="oc-pl-xs"></oc-contextual-helper-stub>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Share" invitelabel="" class="oc-my-s"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" sharedparentroute="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
  <!---->
</div>
`;

exports[`FileShares collaborators list renders sharedWithLabel and sharee list 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm">Share with people</h3>
    <oc-contextual-helper-stub title="Invite specific people" text="Enter a name or group to share this item." list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" endtext="" readmorelink="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" class="oc-pl-xs"></oc-contextual-helper-stub>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Share" invitelabel="" class="oc-my-s"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
  </ul>
  <div class="oc-flex oc-flex-center">
    <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="toggle-shares-list-btn">Show less</oc-button-stub>
  </div>
  <!---->
</div>
`;

exports[`FileShares current space loads space members if a space is given and the current user is member 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm">Share with people</h3>
    <oc-contextual-helper-stub title="Invite specific people" text="Enter a name or group to share this item." list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" endtext="" readmorelink="https://doc.owncloud.com/webui/next/owncloud_web/web_for_users.html#sharing" class="oc-pl-xs"></oc-contextual-helper-stub>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Share" invitelabel="" class="oc-my-s"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-mb-l">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
  <h4 class="oc-text-bold oc-my-s">Space members</h4>
  <ul id="space-collaborators-list" aria-label="Space members" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
</div>
`;
